addons/@mlabs-haskell/godot-cardano/src/tx_builder.gd

```diff
@@ -38,87 +73,120 @@ class MintToken:
 	minting_policy: PlutusScript,
 	tokens: Array[MintToken],
 	redeemer: Object
-) -> void:
+) -> TxBuilder:
 	if !redeemer.has_method("to_data"):
-		push_error("Provided redeemer does not implement `to_data`")
-		return
-		
-	var encoded_redeemer: Cbor.SerializeResult = Cbor.serialize(redeemer.to_data(true), true)
+		_results.push_back(
+			Result.Err.new("Provided redeemer does not implement `to_data`")
+		)
+		return self
+
+	var serialize_result: Cbor.SerializeResult = Cbor.serialize(redeemer.to_data(true), true)
 	
-	if encoded_redeemer.is_err():
-		push_error("Encoding datum failed")
-		return
+	_results.push_back(serialize_result)
+	if serialize_result.is_err():
+		return self
 		
 	var tokens_dict: Dictionary = {}
 	tokens.map(
 		func (token: MintToken) -> void:
 			var prev = tokens_dict.get(token._token_name, BigInt.zero()._b)
 			tokens_dict[token._token_name] = prev.add(token._quantity._b)
 	)
-	_builder.mint_assets(
-		minting_policy,
-		tokens_dict,
-		encoded_redeemer.value
+	
+	var result := Result.VariantResult.new(
+		_builder._mint_assets(
+			minting_policy,
+			tokens_dict,
+			serialize_result.value
+		)
 	)
+	
+	_results.push_back(result)
+	
+	return self
 
 func collect_from(utxos: Array[Utxo]) -> void:
 	_builder.collect_from(utxos)
-
-func complete() -> TxComplete:
-	var wallet_utxos: Array[_Utxo] = []
-	var change_address := _cardano.wallet._get_change_address()
-	var additional_utxos: Array[_Utxo] = []
 	
-	wallet_utxos.assign(
+func complete() -> Result:
+	var wallet_utxos: Array[Utxo] = _cardano.wallet._get_utxos()
+	var _wallet_utxos: Array[_Utxo] = []
+	var change_address := _cardano.wallet._get_change_address()
+	_wallet_utxos.assign(
 		_cardano.wallet._get_utxos().map(func (utxo: Utxo) -> _Utxo: return utxo._utxo)
 	)
+	var additional_utxos: Array[Utxo] = [] # TODO
+	additional_utxos.assign(wallet_utxos)
+	
+	var balance_result := \
+		BalanceResult.new(
+			_builder._balance_and_assemble(_wallet_utxos, change_address._address)
+		)
 	
-	var eval_result := \
-		_builder \
-			.balance_and_assemble(wallet_utxos, change_address._address) \
-			.evaluate(wallet_utxos + additional_utxos)
+	var error = _results.any(func (result: Result) -> bool: return result.is_err())
+	
+	_results.push_back(balance_result)
+	if not error and balance_result.is_ok():
+		var eval_result := balance_result.value.evaluate(wallet_utxos + additional_utxos)
 		
-	return TxComplete.new(
-		_cardano,
-		Transaction.new(
-			_builder.complete(
-				wallet_utxos,
-				change_address._address,
-				eval_result
+		_results.push_back(eval_result)
+		if eval_result.is_ok():
+			return CompleteResult.new(
+				_cardano,
+				_builder._complete(
+					_wallet_utxos,
+					change_address._address,
+					eval_result.value
+				)
 			)
-		)
+	
+	for result in _results:
+		if result.is_err():
+			push_error(result.error)
+			
+	return CompleteResult.new(
+		_cardano,
+		_Result.err("Failed to complete transaction; errors logged to output", 1)
```

Comment:
Does it make sense to return `1` here? A user might think this means `TxBuilderStatus.BAD_PROTOCOL_PARAMETERS` . Maybe  we should add another error variant to `TxBuilderStatus` after`TxBuilderStatus.OTHER_ERROR`  for this. I'm thinking `TxBuilderStatus.COMPLETE_FAILED`.